Allow PyTree inputs for coil objectives

singh-jaydeep opened the pull request:
Adds new input formats for objectives inheriting from ```_CoilObjective```. In particular, 
 
- ```bounds```, ```target```, and ```weight``` can accept pytrees with values for individual coils, provided the input is a prefix of the input ```CoilSet```. weights can be zero, in which case the corresponding coil is masked.
- ~~A new option ```indices``` accepts pytrees with the same structure, indicating whether the bound/target should be applied to given coils.~~ 
- The attribute ```_broadcast_input``` is required for coil objectives, stating if they are computed per-coil or per-grid node.


Notes
- Resolves #1152. 
- This may not be the cleanest/most general approach, and am open to any suggestions! 
- When the user gives an input which is not just a simple scalar we broadcast the input to be of size dim_f and store it. This isn't very efficient, but saves us from adding new logic to the compute function in ```_Objective```.
- The setter functions in ```_Objective``` will have issues with pytree inputs. Would it make sense to add specific setters in ```_CoilObjective```? [resolved]
- Would require tests, etc. [done]

singh-jaydeep committed:
Add PyTree Support for Coil Objectives

- imports jax.tree_util.tree_broadcast in DESC backend
- Allows for "target, bounds, weight" parameters in _CoilObjective to be given as pytrees
- New parameter option "indices" to only compute error on a subset of coils in a MixedCoilSet
- Adds attribute "_mask" to _CoilObjective containing array of 0's (for coils not optimized) and 1's (for those that are)
- Adds conditional to compute_unscaled, computed_scaled, and compute_scaled_error in _Objective which applies the "_mask" array when applicable.

singh-jaydeep committed:
Add PyTree Support for Coil Objectives

- imports jax.tree_util.tree_broadcast in DESC backend
- Allows for "target, bounds, weight" parameters in _CoilObjective to be given as pytrees
- New parameter option "indices" to only compute error on a subset of coils in a MixedCoilSet
- Adds attribute "_mask" to _CoilObjective containing array of 0's (for coils not optimized) and 1's (for those that are)
- Adds conditional to compute_unscaled, computed_scaled, and compute_scaled_error in _Objective which applies the "_mask" array when applicable.

singh-jaydeep committed:
Add PyTree Support for Coil Objectives

- imports jax.tree_util.tree_broadcast in DESC backend
- Allows for "target, bounds, weight" parameters in _CoilObjective to be given as pytrees
- New parameter option "indices" to only compute error on a subset of coils in a MixedCoilSet
- Adds attribute "_mask" to _CoilObjective containing array of 0's (for coils not optimized) and 1's (for those that are)
- Adds conditional to compute_unscaled, computed_scaled, and compute_scaled_error in _Objective which applies the "_mask" array when applicable.

singh-jaydeep committed:
Merge remote-tracking branch 'origin/coil-input-reformat' into coil-input-reformat

singh-jaydeep committed:
Reformat

singh-jaydeep committed:
Merge branch 'PlasmaControl:master' into coil-input-reformat

singh-jaydeep committed:
Add "indices" option to classes inheriting from _CoilObjective

singh-jaydeep committed:
Reverting .gitignore

singh-jaydeep committed:
Add more precise array broadcasting

-Given user pytree input for "weight", "target/bounds", and "indices", broadcasts to larger arrays of size = number of objectives. Was done for objectives which are scalar-valued for single coils, but not vector-valued ones.
- Adds an attribute "_broadcast_input" to classes inheriting from _CoilObjective. Equals "Coil" for objectives which evaluate to a single number per coil (e.g. length), and "Node" for objectives evaluated on individual nodes.

singh-jaydeep committed:
Fix indexing error

YigitElma commented:
@singh-jaydeep is this ready for review?

singh-jaydeep commented:
I'll have it ready by Monday. There's one thing related to memory usage I wanted to check.

singh-jaydeep committed:
Update changelog and improve broadcasting

singh-jaydeep committed:
Merge remote-tracking branch 'origin/coil-input-reformat' into coil-input-reformat

singh-jaydeep committed:
Add indices option to other objectives

dpanici committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep committed:
Clean

dpanici commented on desc/objectives/_coils.py:
```diff
@@ -137,6 +161,31 @@ def _prune_coilset_tree(coilset):
             "Only use toroidal resolution for coil grids.",
         )
 
+        def broadcast_input(arr):
```
It might make more sense to have this as a (maybe private) method of this objective, then when you make the target setter, you can reference this as well.

dpanici commented:
> * The setter functions in `_Objective` will have issues with pytree inputs. Would it make sense to add specific setters in `_CoilObjective`?

Actually question here: if we always in the `_CoilObjective` clean up the target, bounds, etc before we set these attributes, we don't even need to worry about a target setter, right? If that is the case, I'd prefer to then not need the target setter. maybe I am missing something key here though on why we need the extra target setter. I guess it would mainly be to handle the case where an objective is created, and then its target is updated post-hoc?

singh-jaydeep committed:
Add unit tests

- Two tests in test_objective_funs.py. One checks that user input in "bounds", "target", "weight", and "indices" broadcasts correctly. The other checks that the scaled error is correctly masked given "indices".
- One test in test_examples.py. Takes pytree input for "target" and "indices", optimizes, and verifies that output is as desired.

singh-jaydeep committed:
Merge remote-tracking branch 'origin/coil-input-reformat' into coil-input-reformat

singh-jaydeep committed:
Simplify broadcasting operation

singh-jaydeep committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep committed:
Increase codecov (hopefully)

singh-jaydeep committed:
Merge remote-tracking branch 'origin/coil-input-reformat' into coil-input-reformat

YigitElma committed:
Merge branch 'master' into coil-input-reformat

YigitElma committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep committed:
Update docstring for indices

singh-jaydeep commented on desc/objectives/_coils.py:
```diff
@@ -38,7 +40,27 @@ class _CoilObjective(_Objective):
+        Boolean indicating whether particular coils in a CoilSet
+        should be optimized according to the objective. Default
+        is to optimize all coils. If a list, must have the same structure
+        as coil.
```
Sorry I missed this! Will do.

dpanici reviewed (changes requested):
last thing I think I want for this PR: adding an example use of this to the [coil tutorial notebook](https://github.com/PlasmaControl/DESC/blob/master/docs/notebooks/tutorials/coil_stage_two_optimization.ipynb). In that notebook as one of the last cells, we have an optimization of a MixedCoilSet of modular coils and saddle coils. It would be nice to say have the length penalty only target the modular coils (as it effectively does now) and then maybe add some other penalty for the saddle coils, and have this take advantage of the new API as well as show the use of `indices`

And also while we are changing that notebook, adding to the constraints for that last finite beta coil optimization this constraint
`LinkingCurrentConsistency(eq,coilset,eq_fixed=True)` is something I've been meaning to add, and this would be a good spot to change that

@YigitElma do not change the coil notebook on #1877 until this PR is in to avoid annoying conflicts

singh-jaydeep committed:
Add new coil objective options to stage 2 optimization notebook

singh-jaydeep committed:
Add new coil objective options to stage 2 optimization notebook

singh-jaydeep committed:
Update coil opt notebook and test examples

- Tries to reduce memory usage in notebook
- changes test_coil_individual_targets_optimization to verify that the optimized coils have the prescribed targets, and that the gradient with respect to unoptimized coils is zero.

singh-jaydeep committed:
Fix colors

dpanici reviewed (changes requested):
my last change request I promise... I think either leave a comment where mask is used in objective funs to describe why it is there, or add a line about mask to the docstring of _Objective

singh-jaydeep commented:
The "mask" option is proving to be a bit clunky. In addition to getting the normalization somewhat wrong, it doesn't work well for scalar valued loss functions unless the mask gets passed in (e.g. the mean by default includes the masked values). Fixing these things would further clutter up the objective compute functions, which I don't want to do unless masking is useful for other objectives as well. Is there a preferred way of handling this? 

(Another option is if "indices" gets passed into a coilobjective, when building it changes self.things to only include the coils which are being optimized. This is cleaner in that it keeps the masking within the coil objective only. But the optimizer objects when passing in a coilset and objective with "indices", since the set of optimizable things doesn't match.)

singh-jaydeep committed:
Eliminate indices, allow weights to vanish

singh-jaydeep committed:
Test updated coil objectives, update changelog, update stage two optimization notebook

singh-jaydeep committed:
Resolve merge conflict in notebook

singh-jaydeep committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep committed:
Fix determination of dim_f

singh-jaydeep committed:
Merge remote-tracking branch 'origin/coil-input-reformat' into coil-input-reformat

singh-jaydeep committed:
Continue fixing dim_f

YigitElma commented on desc/objectives/_coils.py:
```diff
@@ -38,10 +40,30 @@ class _CoilObjective(_Objective):
         Collocation grid containing the nodes to evaluate at.
         If a list, must have the same structure as coil.
-
+    target : float, list, optional
```
Once you update that function all the coil objective docs will be updated automatically.

YigitElma reviewed (commented):
Besides couple places that need change, it looks good to me. Again, thanks for the contribution!

singh-jaydeep committed:
Cosmetic changes

singh-jaydeep committed:
Remove coil structure utility function

singh-jaydeep committed:
Update collect_docs() logic and cosmetic changes

singh-jaydeep committed:
Remove reference to utility function

singh-jaydeep committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep commented on desc/objectives/objective_funs.py:
```diff
@@ -1287,7 +1288,8 @@ def compute_scaled_error(self, *args, **kwargs):
         if self._loss_function is not None:
             f = self._loss_function(f)
-        return jnp.atleast_1d(self._scale(self._shift(f), **kwargs))
+        f = jnp.atleast_1d(self._scale(self._shift(f), **kwargs))
```
Done

YigitElma reviewed (dismissed):
It looks alright to me. I don't think the notebook result is related to this PR, it has been a while since we last ran that notebook, it might be due to some optimization or other code change.

dpanici committed:
Merge branch 'master' into coil-input-reformat

YigitElma committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep commented:
I installed a fresh venv, and (even on master) can't reproduce the values. Still not sure whether it's architecture differences. I'm on a mac with ARM64. Using JAX 0.6.2 and numpy 2.2.6.

YigitElma committed:
update the notebook

YigitElma committed:
Merge branch 'master' into coil-input-reformat

singh-jaydeep commented on desc/objectives/_coils.py:
```diff
@@ -38,10 +40,30 @@ class _CoilObjective(_Objective):
+    "Coil" if the objective returns a single scalar per coil, and "Node"
+    if it returns a scalar at every grid point. To be compatible with
+    masking, compute function should apply the mask
+    self._coilset_tree["coilset_mask"] before returning data.
```
We got rid of an explicit mask input by allowing weight=0 for some coils, but we still build a mask so that the coilobjective.compute() knows to only return the objective values for unmasked coils. This was to handle loss functions that are min/max/mean. Might be other ways of doing it (e.g. changing logic in Objective class instead).

dpanici committed:
Merge branch 'master' into coil-input-reformat
